### DISCUSSION TITLE
Unable to access useUser hook in _app.tsx

### DISCUSSION THREAD

**Discussion by rdylina** [vercel/next.js#45139] [Help] [answered]
For some reason, maybe this is because I don't fully understand the lifecycle of NextJS at the _app.tsx level, I can't access the useUser hook from within _app.tsx. I'd like to conditionally render the login page from here and if not then render the layout. But for some reason even when logged in, the useUser hook returns null. Every component below _app.tsx can properly access this hook, just not _app.tsx. 

Am I doing something wrong or is this by design?

**Comment by 0xsarwagya** [comment]
It is possible that the problem is related to the lifecycle of the _app.tsx component in Next.js. The _app.tsx component is used to wrap the entire application and is executed on the server-side before the initial render. Therefore, any hooks that rely on client-side state may not be available during the initial render.

It is recommended to move the user check logic to a higher-order component or a component that is rendered later in the lifecycle of the application, such as a page component. You can also make use of the getInitialProps method of the _app.tsx component to retrieve the user data on the server-side and then store it in a store like redux or context.

In addition, it is important to make sure that the useUser hook is only called on the client-side and not on the server-side.

Please let me know if you need more help.

**Comment by guitheengineer** [comment, marked as answer]
Easy solution is to create a wrapper component which inside of it consumes useUser and then redirects to the login page or do a conditional to render a different component.

Note that **if you're using context, then your wrapper must be inside of it.**

```tsx
const LoginWrapper = ({ children }: { children: ReactNode }) => {
  const { booleanState } = useUser()

  // if you want to redirect
  const router = useRouter()
  useEffect(() => {
    if (booleanState) {
      router.push('/login')
    }
  }, [booleanState])

  // or if you want to just render another component
  if (booleanState) {
    return <LoginPage />
  }

  return <>{children}</>
}

export default function App({ Component, pageProps }: AppProps) {
  return (
    <UserContext>
      <LoginWrapper>
        <Component {...pageProps} />
      </LoginWrapper>
    </UserContext>
  )
}
```

**Reply by rdylina** [reply]
Thank you everyone. When the app folder becomes stable it looks like I won't need the wrapper to achieve my goal but until then I'll make one. 

Thanks again!

**Comment by icyJoseph** [comment]
As other's have said the problem is that a hook that reads context, must be used from within a Provider of that context. `_app` for all intents and purposes, is the top level component that you interact with. 

- You can either write a HoC and `export default myHoc(MyApp)` from `_app`
- Or wrap the return JSX of `_app` with a Provider, and move your logic to within that Provider
- Or do as guilhermefront suggests (variation of the above)
